sql: Add ALTER JOB x OWNER TO y

dt opened the pull request:
Jobs have the notion of an owner like other objects in the system, but previously did not have a mechanism for altering this owner to transfer ownership, particularly to or from a role. This adds ALTER JOB x OWNER TO y, mirroring the similar syntax for adjusting the owner of tables, views and other objects in the system.

Release note (sql change): ALTER JOB ... OWNER TO can now be used to transfer ownership of a job between users/roles.
Epic: none.

dt committed:
sql: Add ALTER JOB x OWNER TO y

Jobs have the notion of an owner like other objects in the system, but
previously did not have a mechanism for altering this owner to
transfer ownership, particularly to or from a role. This adds
ALTER JOB x OWNER TO y, mirroring the similar syntax for adjusting
the owner of tables, views and other objects in the system.

Release note (sql change): ALTER JOB ... OWNER TO can now be used to transfer ownership of a job between users/roles.
Epic: none.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/138139)
<!-- Reviewable:end -->

dt committed:
jobs: members of roles are owners of jobs owned by roles

This mirrors the transitive ownership model used by all other objects in the DB.

Release note: none.
Epic: none.

msbutler reviewed (commented):
er, sorry. only reviewed first commit. hold of on merging one sec.

msbutler commented on pkg/sql/alter_job_owner.go:
```diff
@@ -0,0 +1,168 @@
+
+	// Nobody is allowed to transfer jobs to "node".
+	if owner == username.NodeUserName() {
+		return pgerror.Newf(pgcode.InsufficientPrivilege, "cannot transfer ownership to or from node")
```
I don't think typical cockroach users know about the node user. Perhaps rephrase to "cannot transfer ownership to or from the 'node' user, used by cockroach internally"

dt replied:
You cannot create, or use, or even connect as the `node` user already, so I don't think you'd be trying to transfer anything to or from it anyway unless you'd already decided to venture into internals so I think by the point you get this error, you know you're messing with internals already.

dt replied:
Made it "internal node user"

msbutler commented on pkg/sql/alter_job_owner.go:
```diff
@@ -0,0 +1,168 @@
+	if err != nil {
+		return err
+	}
+	for role := range members {
```
do a map look up instead?

dt replied:
doh

dt replied:
Done.

msbutler commented on pkg/sql/alter_job_owner.go:
```diff
@@ -0,0 +1,168 @@
+		}
+	}
+
+	if forNewOwner {
```
nit: instead of passing a `forNewOwner` flag solely for creating an error, you could instead wrap the errors returned by this helper func?

msbutler replied:
ah, you don't want to wrap the pgerror. that's fine.

msbutler commented on pkg/sql/logictest/testdata/logic_test/owner:
```diff
@@ -142,7 +142,7 @@ REVOKE GRANT OPTION FOR ALL PRIVILEGES ON DATABASE d FROM testuser2
 user testuser2
 
 # testuser2 has ALL privileges, no grant options, and is not a member of a role,
-# so it can't GRANT.
```
new years resolution to eradicate apostrophes?

dt replied:
They mess with the quote/unquote matching in any syntax highlighting that thinks these files are SQL, since `#` isn't a valid SQL comment

msbutler commented on pkg/sql/logictest/testdata/logic_test/owner:
```diff
@@ -255,3 +255,121 @@ user testuser
+
+user root
+
+skipif config local-mixed-24.3
```
nit: could you instead declare a subtest that gets skipped in the mixed version config, to avoid all of these `skipif config local-mixed-24.3` lines?

dt replied:
you can't.

msbutler replied:
yeah, perhaps it's best not to mess with logic.go

msbutler commented on pkg/sql/logictest/testdata/logic_test/owner:
```diff
@@ -255,3 +255,121 @@ user testuser
+
+skipif config local-mixed-24.3
+statement ok
+ALTER JOB $testuser_job_id OWNER TO testuser
```
wait, why can we transfer ownership to `testuser`? The job is currently owned by `testuser2` which is apart of the `testrole`, but `testuser` is not a member of this role.

dt replied:
this is where reading a diff is harder than reading the actual file: in the existing test setup, before this added section, on line 152 we do `GRANT testuser TO testuser2;` so testuser2 is a member of the testuser role already.

msbutler replied:
hrm. this test would probably be cleaner in its own file, but i also understand that you're on a crusade to minimize test server inits.

dt replied:
I donno, the file is all about ownership both via direct and via roles, and is only a couple hundred lines, so this feels like where this belongs doesn't it? also, yeah, not worth an extra test server startup.

msbutler reviewed (changes requested):
left a few comments on second patch

dt commented:
TFTR!

bors r+
